### DISCUSSION TITLE
提案4：Starcoin Move 合约标准库升级到 V6 版本

### DISCUSSION THREAD

**Discussion by nanne007** [starcoinorg/starcoin#2806] [DaoProposal]
# 提案4：将 Starcoin Move 合约标准库升级到 V6 版本

本提案提议将 Move 合约标准库（stdlib）升级到 V6 版本，主要包含以下特性：

- 实现 Oracle 协议，支持通用的数据预言机以及价格预言机。 (#2732)
- 实现通用的 NFT 协议。(#2688, #2763, #2760, #2767, #2769, #2771, #2772)
- 实现合约账户（目前只对 stdlib 开放，计划下一步对外开放）。(#2673)
- `Account::deposit` 支持存入 zero 。 (#2745)
- `Account` 支持自动收款功能，收款人不需要事先  `accept_token`。 (#2746)
- 添加了许多脚本函数，供外部调用。 (#2745, #2781)
- 修复 `Math.mul_div` 的 bug。 (#2775)

##升级提案信息

- 升级包二进制：https://github.com/starcoinorg/starcoin/tree/master/vm/stdlib/compiled/6/5-6/stdlib.blob
- Package 哈希：0x20afc123648ff1e2ef4d4339003e9c554a3c4cc9dd1059bdb08e7d7e01d77abfd7
- 提案交易: https://stcscan.io/main/transactions/detail/0x3a652d0b9c396f72a460539978be3a68c7456155317f9faaa2531201472f80be
- 提案发起人：[0xb4491fd2d00db301b8ec6ddb6d5ae747](https://stcscan.io/main/address/0xb4491fd2d00db301b8ec6ddb6d5ae747)
- 提案 id: 4

## 升级流程

1.  从今天发起升级提案开始，进行为期 7 天的投票。
2. 7 天后如果投票通过，则进入公示期，公示期 1 天。如果未通过，则提案失败。
3. 公示期过后，任何人都可以发起升级合约交易（升级交易的 Package hash 必须等于提案中的 Package hash），将标准库的 v6 版本写入到链上。该升级交易执行后，新特性激活。 **由于新特性包含了硬分叉特性，届时尚未升级到 starcoin v1.5.0 版本的节点可能会和主网分叉。**

### 关于 Starcoin 治理的更多信息请参看：

* [Starcoin 技术白皮书中的治理部分](https://developer.starcoin.org/zh/sips/sip-2/)
* [链上治理的生命周期](https://developer.starcoin.org/zh/key_concepts/dao_governance/)
* [通过治理机制修改 DAO 的设置](https://developer.starcoin.org/zh/cli/modify_dao_config/)

当前，投票治理的 DApp 尚在开发中，需要通过 CLI 命令进行投票。

注意：投票时，会将当前账户的 STC 抵押到合约中，直到投票结束，当前投票周期为 7 天。期间主动退款则视为放弃投票。** 为了鼓励参与链上治理，未来会推出链上治理的激励措施。本次参与投票的用户会根据激励策略进行补偿。**

### 准备
1. 下载 starcoin 版本 v1.5.0，并同步主网的区块，进入 starcoin 控制台。
2. 将自己的账户导入到节点内置的钱包中，或者从交易所提币到节点钱包的地址中。

### 操作命令

1. 查看投票状态

```
dev call --function 0x1::Dao::proposal_state -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747 --arg 4
```
第一个参数是提案发起人的地址，第二个参数是提案 id，返回的状态为 1 表示等待状态， 2 表示正在投票中，4 表示投票通过。

2. 执行投票

通过一下命令进行投票：

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::cast_vote -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747 --arg 4 --arg true --arg 1000000000u128 -b
```

第一个参数是提案发起人的地址，第二个参数是提案 id，第三个参数通过 true/false 表示同意和反对，第四个参数表示用来投票的 STC 数量，注意单位是 nanoSTC，有9位是小数点后的单位，上面的例子是 1 STC。

注意：替换 your_address 为你自己的地址，不要用全部的 STC，要留够 gas 费。

3. 放弃投票

如果想放弃投票，取回抵押的 STC，可以用以下命令：

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::revoke_vote -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747 --arg 4 -b
```

4. 投票结束后取回抵押的 STC

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::unstake_vote -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747 --arg 4 -b
```

5. 查看投票结果

```
dev call --function 0x1::Dao::proposal_info -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747

{
  "ok": [
    4,
    1629912160339,
    1630516960339,
    0,
    0
  ]
}
```

返回值的第一个数表示提案 id，第二个数是开始时间，第三个数是结束时间，第四个数是投票同意的票数，第五个数是投票反对的票数。

--------------
## English Guide

Voting for the Starcoin Move stdlib v6 upgrade is open. This upgrade is the second stdlib upgrade since the Starcoin mainnet lanuched and includes the following features.

- Implement Oracle protocol, support general data oracle and price oracles. (#2732)
- Implement NFT protocol which has builtin IdentifierNFT, GenesisNFT. (#2688, #2763, #2760, #2767, #2769, #2771, #2772)
- Support contract account by using `SignerCapability` abstraction. (#2673)
- `Account::deposit` support deposit zero token. (#2745)
- `Account` supports auto-accept-token feature. (#2746)
- Add many script functions. (#2745, #2781)
- Fix `Math.mul_div`. (#2775)

## Upgrade proposal information

- Upgrade package binary: https://github.com/starcoinorg/starcoin/tree/master/vm/stdlib/compiled/6/5-6/stdlib.blob
- Package hash: 0x20afc123648ff1e2ef4d4339003e9c554a3c4cc9dd1059bdb08e7d7e01d77abfd7
- Proposal transaction: https://stcscan.io/main/transactions/detail/0x3a652d0b9c396f72a460539978be3a68c7456155317f9faaa2531201472f80be
- Proposer address: [0xb4491fd2d00db301b8ec6ddb6d5ae747](https://stcscan.io/main/address/0xb4491fd2d00db301b8ec6ddb6d5ae747)
- Proposal id: 4

Coin holders are invited to vote on the chain. When voting, the STC of the current account will be stake to the contract until the end of the voting period, which is currently 7 days.  ** To encourage participation in on-chain governance, incentives for on-chain governance will be introduced in the future. Users who participate in voting this time will be compensated according to the incentive strategy. **

### Preparation

1. Download the  starcoin v1.5.0 and join the mainnet.
2. Import your account into the node's built-in wallet or withdraw coins from the exchange to the node's wallet address.

### Operation commands

1. Check the voting status

```
dev call --function 0x1::Dao::proposal_state -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747 --arg 4
```
The first argument is the address of the proposer, the second argument is the proposal id, and the returned status is 1 for pending status, 2 for voting, and 4 for voting passed.

2. Execute voting

Voting is performed by the following command.

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::cast_vote -t 0x1::STC::STC -t 0x1:: UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747 --arg 4 --arg true --arg 1000000000u128 -b
```

The first argument is the proposer, the second argument is the proposal id, the third argument indicates agreement and disagreement via true/false, and the fourth argument indicates the number of STCs used to vote, note that the unit is nanoSTC with 9 decimal places, the above example is 1 STC.

Note: Replace your_address with your own address, do not use all the STCs, but leave enough for the gas fee.

3. Revoke Voting

If you want to abort the poll and get back the STCs, you can use the following command.

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::revoke_vote -t 0x1::STC::STC -t 0x1:: UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747 --arg 4 -b
```

4. retrieve the staked STC after the poll

```
account execute-function -s your_address --function 0x1::DaoVoteScripts::unstake_vote -t 0x1::STC::STC -t 0x1:: UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747 --arg 4 -b
```

5. View poll results

```
dev call --function 0x1::Dao::proposal_info -t 0x1::STC::STC -t 0x1::UpgradeModuleDaoProposal::UpgradeModuleV2 --arg 0xb4491fd2d00db301b8ec6ddb6d5ae747

{
  "ok": [
    4,
    1629912160339,
    1630516960339,
    0,
    0
  ]
}
```

The first number of the return value represents the proposal id, the second number is the start time, the third number is the end time, the fourth number is the number of votes for and the fifth number is the number of votes against.

**Comment by timyangnet** [comment]
PS: This is the same vote as id 3 last week. Because of the end time web page display bug, some voters missed the vote 3 deadline, so the proposer reopened the proposal again.

PS：投票4 和id 3的投票是一样的。 因为结束时间的网页显示错误，有些投票者错过了投票3的最后期限，所以提议者又重新开始了提案。
